docs/design/2021-12-08-instance-scope.md

```diff
@@ -0,0 +1,224 @@
+
+```
+# tidb.toml
+[instance]
+max_connections = 1234
+tidb_general_log = "/path/to/file"
+```
+
+The default will be that we map variables as read-only, so we don't need to think about the specific cases (i.e. you can't easily change the socket or port). This helps us achieve completeness first, and then we can then evaluate which ones can be made dynamic.
+
+### Stage 3: Refactoring
+
+The use of a `GetGlobal()` and `SetGlobal()` func for each instance scoped system variable is not ideal. It is possible to refactor the system variable framework so that instance scope is stored in a map, and the values are updated automatically by `SET GLOBAL` on an instance scoped variable. On startup, as the configuration file is parsed it will update the values in the map. This seems like a better approach than the current use of Setters/Getters, and because there is a prescribed way of doing it we can correctly handle the data races that are common with our current incorrect usage of calling `config.GetGlobalConfig()`.
+
+Thus, the source of truth for instance scoped variables moves from the `config` package to another part of the server (likely `domain`). See [issue #30366](https://github.com/pingcap/tidb/issues/30366).
+
+## Impacts & Risks
+
+The biggest risk is that we can not agree on a reduced scope of implementation and the project extends to cover increased scope. Configuration management is a classic example of a [bikeshed problem](https://en.wikipedia.org/wiki/Law_of_triviality), and we will likely need to make some tradeoffs to get anywhere.
+
+Many of the alternatives are difficult to implement because they break compatibility (we need to live with `GLOBAL` means cluster `GLOBAL`, and there is no `SET CLUSTER`) or they break rolling upgrade scenarios because we currently have no specific rules of the minimum version of TiDB which can be upgraded from.
+
+## Investigation & Alternatives
+
+- There is no equivalent functionality to reference in MySQL since it does not have the native concept of a cluster.
+- An [earlier proposal](https://docs.google.com/document/d/1RuajYAFVsjJCwCBpIYF--9jtgxDZZcz3cbP-iVuLxJI/edit) for `INSTANCE` scope, with rules for precedence (same author)
+- CockroachDB has cluster level settings and node-level settings. Most settings are cluster level, and [node-level](https://www.cockroachlabs.com/docs/v21.2/cockroach-start) needs to be parsed as arguments when starting the server.
+
+## Unresolved Questions
+
+There are **very few** known scenarios where a system variable is *currently both* instance and global scoped, but these individually would need to be handled:
```

Comment:
I still don't get it: why do we replace `global` with `instance` instead of replacing `session` with `instance`? To users, they are both hard to understand and you need to explain it in the user docs anyway.
Are there any cases where a system variable is both session and instance scoped? I don't think so.

Reply:
That's correct, there are no there are no cases where a variable is both session and instance.

Session by semantics is supposed to be thread-local-storage. When the initial connection is created to MySQL it will copy all the values of global variables that also have session scope, and any further changes to global will not affect those session changes. It should only be changes that you make in your connection that change the value.

What I am calling `INSTANCE` is actually the identical semantics to MySQL `GLOBAL`. They are much cluser in behavior: the only difference is you will need to check the docs to see if it is an INSTANCE-GLOBAL or a GLOBAL-GLOBAL if you want to know if your change (1) propagates to the cluster and (2) persists.

Reply:
This is listed as motivation #3 in "Motivation and Background".

Reply:
I know the semantics between `session` and `instance` is different, but the semantics between `instance` and `global` is also different. From users' view:
- If you set a global system variable and then reboot TiDB, the setting should still take effect, but it actually does not.
- If you set a global system variable and then connect to TiDB from another session, the setting should take effect, but it may not.
So I don't think `instance` is closer to `session`.

Reply:
> If you set a global system variable and then connect to TiDB from another session, the setting should take effect, but it may not.

This is not the case since ~TiDB 5.2. On a single-instance basis if you set a `GLOBAL `system variable and then connect from another session it will take effect unless the scope is `SESSION + GLOBAL`, and the other session pre-originated the setting. On a multi-instance basis there is an etcd notification sent immediately, but there is a race where you could beat it.

> So I don't think instance is closer to session.

I was arguing it is closer to MySQL `GLOBAL`, or rather it is identical. TiDB `GLOBAL` does not exist in MySQL.

Reply:
I have added a section on "Documentation Changes", PTAL